dice-game/service/src/main.rs

```diff
@@ -0,0 +1,139 @@
+use rand::Rng;
+use map_vec::Map;
+use oasis_std::{Address, Context};
+
+#[derive(oasis_std::Service)]
+struct DiceGame {
+    max_score: u32,
+    num_players: u32,
+    scores: Map<Address, u32>,
+}
+
+type Result<T> = std::result::Result<T, String>;
+impl DiceGame {
+    pub fn new(_ctx: &Context, num_players: u32) -> Self {
+        Self {
+            max_score: 0,
+            num_players,
+            scores: Map::new(),
+        }
+    }
+
+    /// returns whether still in play
+    pub fn is_in_play(&self, _ctx: &Context) -> bool {
+        self.scores.len() < (self.num_players as usize)
+    }
+
+    /// returns num_players 
+    pub fn num_players(&self, _ctx: &Context) -> u32 {
+        self.num_players
+    }
+
+    /// returns the max score in play
+    pub fn max_score(&self, _ctx: &Context) -> u32 {
+        self.max_score
+    }
+
+    /// rolls a dice, resulting in a random number from 1-6
+    pub fn roll(&mut self, ctx: &Context) -> Result<()> {
+        if self.scores.len() >= (self.num_players as usize) {
+            return Err("Maximum number of players rolled".to_string());
+        }
+
+        // each player is only allowed to roll once
+        if self.scores.contains_key(&ctx.sender()) {
+            return Err("Repeat rolls not allowed!".to_string());
+        }
+
+        let score = rand::thread_rng().gen_range(1, 7);
+        if score > self.max_score {
+            self.max_score = score;
+        }
+
+        self.scores.insert(ctx.sender(), score);
+        Ok(())
+    }
+
+    /// A magic roll that returns a specified value. Used for testing purposes
+    pub fn magic_roll(&mut self, ctx: &Context, value: u32) -> Result<()> {
```

Comment:
How about you have 
```rust
pub fn roll(&mut self, ctx: &Context) -> Result<()>
```

return the `score` and use it to track the winner in your tests instead of using this method?

Reply:
https://github.com/oasislabs/tutorials/pull/56